### DISCUSSION TITLE
Deserialize a null to DateTime

### DISCUSSION THREAD

**Discussion by hantsy** [symfony/symfony#45398] [Q&A] [answered]
The sample project is based on Symfony 6 and PHP 8.1.

I create a web test to verify the */posts/{id}* endpoint, the result is a JSON string for a `Post` object, but the `publishedAt` is a `null`.

```php
$client->jsonRequest('GET', $url); 
$getByIdResponse = $client->getResponse();
echo("json response:::" . $getByIdResponse->getContent());
```

I tried to deserialize the response content to an Post object.  

```php
$getData = $this->getContainer()->get('serializer')->deserialize($getByIdResponse->getContent(), Post::class, "json");
```
But it threw an exception,

```bash 
Symfony\Component\Serializer\Exception\NotNormalizableValueException : 
Failed to denormalize attribute "publishedAt" value for class "App\Entity\Post": 
Expected argument of type "DateTime", "null" given at property path "publishedAt".
```

The `publishedAt` in the Post is a  `DateTime|null` mixed type.

The [sample codes](https://github.com/hantsy/symfony-sample/blob/enum/rest-sample/tests/Controller/PostControllerTest.php#L63-L70) is available on my Github account.

**Comment by chyshyya** [comment, marked as answer]
With 
```
    /**
     * @param DateTime $publishedAt
     */
    public function setPublishedAt(DateTime $publishedAt): self
    {
        $this->publishedAt = $publishedAt;
        return $this;
    }
```
null value not allowed
It must be:
```
    public function setPublishedAt(?DateTime $publishedAt): self
```
? is essential

**Reply by chyshyya** [reply]
And if you want serializer work propertly and automatically with datetime use : \DateTimeInterface instead of \DateTime in entity

**Reply by hantsy** [reply]
Thanks for your suggestion.  Updated, and it works now. 

Is there possible to configure globally to skip all null and empty fields(string, array, collection) of an object when serializing an object to response?

**Reply by chyshyya** [reply]
I didn't try, but this can do the job:
```
# config/packages/framework.yaml
framework:
    # ...
    serializer:
        default_context:
            skip_null_values: true
```
Tell if it will work for you.

**Reply by chyshyya** [reply]
And you can do anything you want as we do with:
https://symfony.com/doc/current/serializer/custom_normalizer.html

**Reply by chyshyya** [reply]
In normalize method you can unset any null or empty normalized keys

**Reply by hantsy** [reply]
Thanks @chyshyya
